### PR TITLE
New Hud buttons

### PR DESCRIPTION
**Pull request by nnngrach**
https://github.com/osmandapp/OsmAnd-Issues/issues/389

Screenshot 2021-07-21 at 17 35 38
Screenshot 2021-07-21 at 17 38 14

video:
https://www.dropbox.com/s/fj9mrjkegd9i3m4/Screen%20Recording%202021-07-22%20at%2000.10.37.mov?dl=0

https://www.dropbox.com/s/2z0h41ho7wtq1g1/Screen%20Recording%202021-07-22%20at%2000.40.15.mov?dl=0

### CODE REVIEW

File: Sources/Controllers/2.0/OAQuickActionHudViewController.mm [hunk 1]
```diff
@@ -55,9 +58,13 @@ - (void)viewDidLoad
 {
     [super viewDidLoad];
     
-    [_quickActionFloatingButton setImage:[UIImage templateImageNamed:@"ic_custom_quick_action"]
-                                forState:UIControlStateNormal];
-    _quickActionFloatingButton.hidden = ![_settings.quickActionIsOn get];
+    _quickActionFloatingButton.alpha = [_settings.quickActionIsOn get] ? 1 : 0;
+    _quickActionFloatingButton.userInteractionEnabled = [_settings.quickActionIsOn get];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OAQuickActionHudViewController.mm:62]: Test is the area behind the hidden quick action button can be used for map interaction.

File: Sources/Controllers/2.0/OAQuickActionHudViewController.mm [hunk 2]
```diff
@@ -55,9 +58,13 @@ - (void)viewDidLoad
...
+    _quickActionFloatingButton.tintColorDay =  UIColorFromRGB(color_primary_purple);
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OAQuickActionHudViewController.mm:63]: Too many spaces

File: Sources/Controllers/2.0/OATargetMenuViewController.mm [hunk 3]
```diff
@@ -836,14 +837,21 @@ - (void) setMiddleToolbarAlpha:(CGFloat)alpha
 {
     if ([self hasTopToolbar])
     {
-        CGFloat backButtonAlpha = alpha;
+        CGFloat backButtonAlpha = alpha * 2;
+        backButtonAlpha = backButtonAlpha > 1 ? 1 : backButtonAlpha;
+        
+
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OATargetMenuViewController.mm:843]: Extra newline

File: Sources/Controllers/2.0/OAQuickActionHudViewController.mm [hunk 4]
```diff
@@ -150,13 +183,13 @@ - (void) setQuickActionButtonPosition
     {
         if (isLandscape)
         {
-            x = _mapHudController.mapModeButton.frame.origin.x - w;
+            x = _mapHudController.mapModeButton.frame.origin.x - w - 16.0;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OAQuickActionHudViewController.mm:186]: Extract to constants

File: Sources/Controllers/Map/OAMapHudViewController.mm [hunk 5]
```diff
@@ -973,6 +1055,23 @@ - (BOOL) topControlsVisible
     return _statusBarView.alpha > 0;
 }
 
+- (void) setTopControlsAlpha:(CGFloat)alpha
+{
+    CGFloat alphaEx = self.contextMenuMode ? 0.0 : alpha;
```
Comment (PaulStets) [inline Sources/Controllers/Map/OAMapHudViewController.mm:1060]: Will this leave top controls active?

File: Sources/Controllers/Map/OAMapHudViewController.mm [hunk 6]
```diff
@@ -1085,10 +1200,13 @@ - (void) hideBottomControls:(CGFloat)menuHeight animated:(BOOL)animated
...
             _driveModeButton.alpha = 0.0;
             
-            _optionsMenuButton.frame = CGRectMake(0.0, DeviceScreenHeight - 63.0 - menuHeight - bottomMargin, _optionsMenuButton.bounds.size.width, _optionsMenuButton.bounds.size.height);
-            _driveModeButton.frame = CGRectMake(57.0, DeviceScreenHeight - 63.0 - menuHeight - bottomMargin, _driveModeButton.bounds.size.width, _driveModeButton.bounds.size.height);
-            _mapModeButton.frame = CGRectMake(DeviceScreenWidth - 128.0, DeviceScreenHeight - 69.0 - menuHeight - bottomMargin, _mapModeButton.bounds.size.width, _mapModeButton.bounds.size.height);
-            _zoomButtonsView.frame = CGRectMake(DeviceScreenWidth - 68.0, DeviceScreenHeight - 129.0 - menuHeight - bottomMargin, _zoomButtonsView.bounds.size.width, _zoomButtonsView.bounds.size.height);
+            [self setupBottomContolMarginsForHeight:menuHeight];
+            
+            CGFloat offsetValue = DeviceScreenWidth;
+            [self addToView:_optionsMenuButton xOffset:-offsetValue yOffset:0];
```
Comment (PaulStets) [inline Sources/Controllers/Map/OAMapHudViewController.mm:1206]: addOffsetToView

File: Sources/Controllers/Panels/OAMapPanelViewController.mm [hunk 7]
```diff
@@ -779,15 +779,37 @@ - (void) closeRouteInfo:(void (^)(void))onComplete
...
 - (CGRect) shadowButtonRect
 {
     return self.view.frame;
 }
 
 - (void) removeGestureRecognizers
 {
-    while (self.view.gestureRecognizers.count > 0)
-        [self.view removeGestureRecognizer:self.view.gestureRecognizers[0]];
+    for (UIGestureRecognizer *recognizer in self.view.gestureRecognizers)
```
Comment (PaulStets) [inline Sources/Controllers/Panels/OAMapPanelViewController.mm:808]: Check if the panel cannot be dragged while other screens (like context menu) are presented

File: Sources/Extensions/OAHudButton.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+//  OAHudButton.h
+//  OsmAnd Maps
+//
+//  Created by nnngrach on 13.07.2021.
+//  Copyright © 2021 OsmAnd. All rights reserved.
+//
+
+#import <UIKit/UIKit.h>
+
+@interface OAHudButton : UIButton
```
Comment (PaulStets) [inline Sources/Extensions/OAHudButton.h:11]: In objective c and swift and extension has a different meaning to a subclass. Please do not put subclasses into a folder called Extensions.

File: Sources/Controllers/2.0/OAQuickActionHudViewController.mm (hunk 1 above)
Comment (nnngrach) [reply Sources/Controllers/2.0/OAQuickActionHudViewController.mm:63]: https://www.dropbox.com/s/zw9kedqfffl6tc7/Screen%20Recording%202021-08-05%20at%2015.38.58.mov?dl=0

File: Sources/Controllers/Map/OAMapHudViewController.mm (hunk 5 above)
Comment (nnngrach) [reply Sources/Controllers/Map/OAMapHudViewController.mm:1060]: https://www.dropbox.com/s/wtmbnp59hsx1qq8/Screen%20Recording%202021-08-05%20at%2015.57.01.mov?dl=0

File: Sources/Controllers/Panels/OAMapPanelViewController.mm (hunk 7 above)
Comment (nnngrach) [reply Sources/Controllers/Panels/OAMapPanelViewController.mm:808]: https://www.dropbox.com/s/amyxyr7hway7x39/Screen%20Recording%202021-08-05%20at%2016.25.19.mov?dl=0

**Pull request merged**